File: app/src/main/java/com/ozkanmut/ilactakip/DoseStateEngine.kt
package com.ozkanmut.ilactakip

import android.content.Context
import java.time.LocalDate

enum class DoseSessionStatus { UNKNOWN, PENDING, SNOOZED, TAKEN, MISSED, CONFLICT }

data class DoseSessionState(
    val time: String,
    val status: DoseSessionStatus,
    val latestEvent: DoseEvent?,
    val medications: List<Medication>,
    val conflictEvents: List<DoseEvent> = emptyList(),
    val scheduledDate: String = LocalDate.now().toString()
)

object DoseStateEngine {
    private val stateTypes = setOf(
        "alarm", "snoozed", "taken", "missed", "undo_taken", "undo_missed",
        "conflict_resolved_taken", "conflict_resolved_missed"
    )

    fun stateForTime(c: Context, time: String, date: LocalDate = LocalDate.now()): DoseSessionState {
        val scheduledDate = date.toString()
        val events = EventStore.load(c).filter { it.time == time && eventDate(it) == scheduledDate }
        val scheduleMeds = Store.load(c).filter { time in it.times && ProgramRuleStore.isActiveOn(c, it.id, date) }
        return reduce(time, events, scheduleMeds, scheduledDate)
    }

    fun today(c: Context): List<DoseSessionState> {
        val date = LocalDate.now()
        val times = (Store.load(c).filter { ProgramRuleStore.isActiveOn(c, it.id, date) }.flatMap { it.times } +
            EventStore.load(c).filter { eventDate(it) == date.toString() }.map { it.time })
            .filter { it.isNotBlank() }.distinct().sorted()
        return times.map { stateForTime(c, it, date) }
    }

    fun unresolved(c: Context): List<DoseSessionState> = today(c).filter {
        it.status == DoseSessionStatus.PENDING || it.status == DoseSessionStatus.SNOOZED || it.status == DoseSessionStatus.CONFLICT
    }

    private fun reduce(time: String, events: List<DoseEvent>, scheduleMeds: List<Medication>, scheduledDate: String): DoseSessionState {
        if (events.isEmpty()) return DoseSessionState(time, DoseSessionStatus.UNKNOWN, null, scheduleMeds, scheduledDate = scheduledDate)
        val sessionEvents = events.filter { it.type in stateTypes }
        if (sessionEvents.isEmpty()) return DoseSessionState(time, DoseSessionStatus.UNKNOWN, events.firstOrNull(), scheduleMeds, scheduledDate = scheduledDate)

        val resolutions = sessionEvents.filter { it.type == "conflict_resolved_taken" || it.type == "conflict_resolved_missed" }
        if (resolutions.isNotEmpty()) {
            val latestPerActor = resolutions.groupBy { it.actorTopic }.values.mapNotNull(::newestForActor)
            val maxRevision = latestPerActor.maxOf { it.revision }
            val top = latestPerActor.filter { it.revision == maxRevision }
            val topTaken = top.filter { it.type == "conflict_resolved_taken" }
            val topMissed = top.filter { it.type == "conflict_resolved_missed" }

            if (topTaken.isNotEmpty() && topMissed.isNotEmpty()) {
                val conflicts = ordered(listOf(topTaken.maxBy { it.eventId }, topMissed.maxBy { it.eventId }))
                return DoseSessionState(time, DoseSessionStatus.CONFLICT, conflicts.last(), medsFrom(conflicts.last(), scheduleMeds), conflicts, scheduledDate)
            }

            val resolution = top.maxWithOrNull(compareBy<DoseEvent> { it.actorTopic }.thenBy { it.eventId })!!
            val laterUndo = sessionEvents.filter { (it.type == "undo_taken" || it.type == "undo_missed") && it.actorTopic == resolution.actorTopic }
                .maxWithOrNull(DoseEventOrder.withinActor)
            if (laterUndo != null && laterUndo.revision > resolution.revision) {
                return DoseSessionState(time, DoseSessionStatus.PENDING, laterUndo, medsFrom(laterUndo, scheduleMeds), scheduledDate = scheduledDate)
            }
            val status = if (resolution.type == "conflict_resolved_taken") DoseSessionStatus.TAKEN else DoseSessionStatus.MISSED
            return DoseSessionState(time, status, resolution, medsFrom(resolution, scheduleMeds), scheduledDate = scheduledDate)
        }

        val latestByActor = sessionEvents.groupBy { it.actorTopic }.mapValues { (_, actorEvents) ->
            newestForActor(actorEvents.filter { it.type == "taken" || it.type == "missed" || it.type == "undo_taken" || it.type == "undo_missed" })
        }.values.filterNotNull()
        val effective = latestByActor.filterNot { it.type == "undo_taken" || it.type == "undo_missed" }
        if (effective.isEmpty() && latestByActor.any { it.type == "undo_taken" || it.type == "undo_missed" }) {
            val undo = latestByActor.maxWithOrNull(DoseEventOrder.global)!!
            return DoseSessionState(time, DoseSessionStatus.PENDING, undo, medsFrom(undo, scheduleMeds), scheduledDate = scheduledDate)
        }

        if (effective.isNotEmpty()) {
            // Revision is a Lamport-style causal clock. Only opposing terminal
            // facts at the same highest logical revision are truly concurrent.
            // A higher revision is a later decision and must settle older facts.
            val maxRevision = effective.maxOf { it.revision }
            val top = effective.filter { it.revision == maxRevision }
            val topTaken = top.filter { it.type == "taken" }
            val topMissed = top.filter { it.type == "missed" }

            if (topTaken.isNotEmpty() && topMissed.isNotEmpty()) {
                val conflicts = ordered(listOf(
                    topTaken.maxWithOrNull(DoseEventOrder.global)!!,
                    topMissed.maxWithOrNull(DoseEventOrder.global)!!
                ))
                return DoseSessionState(time, DoseSessionStatus.CONFLICT, conflicts.last(), medsFrom(conflicts.last(), scheduleMeds), conflicts, scheduledDate)
            }

            val terminal = top.maxWithOrNull(DoseEventOrder.global)!!
            val status = if (terminal.type == "taken") DoseSessionStatus.TAKEN else DoseSessionStatus.MISSED
            return DoseSessionState(time, status, terminal, medsFrom(terminal, scheduleMeds), scheduledDate = scheduledDate)
        }

        val latest = sessionEvents.maxWithOrNull(DoseEventOrder.global)!!
        val status = when (latest.type) {
            "snoozed" -> DoseSessionStatus.SNOOZED
            "alarm" -> DoseSessionStatus.PENDING
            "undo_taken", "undo_missed" -> DoseSessionStatus.PENDING
            else -> DoseSessionStatus.UNKNOWN
        }
        return DoseSessionState(time, status, latest, medsFrom(latest, scheduleMeds), scheduledDate = scheduledDate)
    }

    private fun medsFrom(event: DoseEvent, fallback: List<Medication>): List<Medication> = if (event.medications.isNotEmpty()) event.medications else fallback
    private fun eventDate(event: DoseEvent): String = event.scheduledDate.ifBlank { LocalDate.now().toString() }
    private fun ordered(events: List<DoseEvent>) = events.sortedWith(DoseEventOrder.global)
    private fun newestForActor(events: List<DoseEvent>): DoseEvent? = events.maxWithOrNull(DoseEventOrder.withinActor)
}

File: ios/Dosefolk/Dosefolk/DoseStateEngine.swift
import Foundation

enum DoseSessionStatus: String, Codable, Equatable {
    case unknown, pending, snoozed, taken, missed, conflict
}

struct DoseSessionState: Equatable {
    var time: String
    var status: DoseSessionStatus
    var latestEvent: DoseEvent?
    var medications: [Medication]
    var conflictEvents: [DoseEvent] = []
    var scheduledDate: String
}

struct DoseStateEngine {
    private let store: LocalStore
    private let calendar: Calendar

    init(store: LocalStore, calendar: Calendar = .current) {
        self.store = store
        self.calendar = calendar
    }

    func stateForTime(_ time: String, scheduledDate: String) throws -> DoseSessionState {
        let events = try DoseEventStore(store: store).all().filter {
            $0.time == time && eventDate($0, fallback: scheduledDate) == scheduledDate
        }
        let medications = try store.load([Medication].self, from: .medications, default: [])
            .filter { $0.times.contains(time) }
        return Self.reduce(time: time, events: events, scheduleMedications: medications, scheduledDate: scheduledDate)
    }

    static func reduce(time: String, events: [DoseEvent], scheduleMedications: [Medication], scheduledDate: String) -> DoseSessionState {
        let stateTypes: Set<String> = ["alarm", "snoozed", "taken", "missed", "undo_taken", "undo_missed", "conflict_resolved_taken", "conflict_resolved_missed"]
        guard !events.isEmpty else { return .init(time: time, status: .unknown, latestEvent: nil, medications: scheduleMedications, scheduledDate: scheduledDate) }
        let session = events.filter { stateTypes.contains($0.type) }
        guard !session.isEmpty else {
            let latest = events.max(by: eventLess)
            return .init(time: time, status: .unknown, latestEvent: latest, medications: meds(latest, fallback: scheduleMedications), scheduledDate: scheduledDate)
        }

        let resolutions = session.filter { $0.type == "conflict_resolved_taken" || $0.type == "conflict_resolved_missed" }
        if !resolutions.isEmpty { return reduceResolutions(time: time, session: session, resolutions: resolutions, fallback: scheduleMedications, scheduledDate: scheduledDate) }

        let latestByActor = Dictionary(grouping: session, by: \.actorTopic).values.compactMap { actorEvents in
            actorEvents.filter { ["taken", "missed", "undo_taken", "undo_missed"].contains($0.type) }.max(by: actorLess)
        }
        let effective = latestByActor.filter { $0.type != "undo_taken" && $0.type != "undo_missed" }
        if effective.isEmpty, let undo = latestByActor.filter({ $0.type.hasPrefix("undo_") }).max(by: eventLess) {
            return .init(time: time, status: .pending, latestEvent: undo, medications: meds(undo, fallback: scheduleMedications), scheduledDate: scheduledDate)
        }
        if !effective.isEmpty {
            let taken = effective.filter { $0.type == "taken" }
            let missed = effective.filter { $0.type == "missed" }
            if !taken.isEmpty && !missed.isEmpty {
                let conflicts = [taken.max(by: eventLess)!, missed.max(by: eventLess)!].sorted(by: eventLess)
                let latest = conflicts.last!
                return .init(time: time, status: .conflict, latestEvent: latest, medications: meds(latest, fallback: scheduleMedications), conflictEvents: conflicts, scheduledDate: scheduledDate)
            }
            let terminal = effective.max(by: eventLess)!
            return .init(time: time, status: terminal.type == "taken" ? .taken : .missed, latestEvent: terminal, medications: meds(terminal, fallback: scheduleMedications), scheduledDate: scheduledDate)
        }
        let latest = session.max(by: eventLess)!
        let status: DoseSessionStatus = latest.type == "snoozed" ? .snoozed : (["alarm", "undo_taken", "undo_missed"].contains(latest.type) ? .pending : .unknown)
        return .init(time: time, status: status, latestEvent: latest, medications: meds(latest, fallback: scheduleMedications), scheduledDate: scheduledDate)
    }

    private static func reduceResolutions(time: String, session: [DoseEvent], resolutions: [DoseEvent], fallback: [Medication], scheduledDate: String) -> DoseSessionState {
        let latestPerActor = Dictionary(grouping: resolutions, by: \.actorTopic).values.compactMap { $0.max(by: actorLess) }
        let maxRevision = latestPerActor.map(\.revision).max() ?? 0
        let top = latestPerActor.filter { $0.revision == maxRevision }
        let taken = top.filter { $0.type == "conflict_resolved_taken" }
        let missed = top.filter { $0.type == "conflict_resolved_missed" }
        if !taken.isEmpty && !missed.isEmpty {
            let conflicts = [taken.max { $0.eventId < $1.eventId }!, missed.max { $0.eventId < $1.eventId }!].sorted(by: eventLess)
            let latest = conflicts.last!
            return .init(time: time, status: .conflict, latestEvent: latest, medications: meds(latest, fallback: fallback), conflictEvents: conflicts, scheduledDate: scheduledDate)
        }
        let resolution = top.max { a, b in a.actorTopic == b.actorTopic ? a.eventId < b.eventId : a.actorTopic < b.actorTopic }!
        let laterUndo = session.filter { ($0.type == "undo_taken" || $0.type == "undo_missed") && $0.actorTopic == resolution.actorTopic }.max(by: actorLess)
        if let laterUndo, laterUndo.revision > resolution.revision {
            return .init(time: time, status: .pending, latestEvent: laterUndo, medications: meds(laterUndo, fallback: fallback), scheduledDate: scheduledDate)
        }
        return .init(time: time, status: resolution.type == "conflict_resolved_taken" ? .taken : .missed, latestEvent: resolution, medications: meds(resolution, fallback: fallback), scheduledDate: scheduledDate)
    }

    private func eventDate(_ event: DoseEvent, fallback: String) -> String { event.scheduledDate.isEmpty ? fallback : event.scheduledDate }
    private static func meds(_ event: DoseEvent?, fallback: [Medication]) -> [Medication] { guard let event, !event.medications.isEmpty else { return fallback }; return event.medications }
    private static func actorLess(_ a: DoseEvent, _ b: DoseEvent) -> Bool { a.revision == b.revision ? a.eventId < b.eventId : a.revision < b.revision }
    private static func eventLess(_ a: DoseEvent, _ b: DoseEvent) -> Bool { a.revision == b.revision ? (a.actorTopic == b.actorTopic ? a.eventId < b.eventId : a.actorTopic < b.actorTopic) : a.revision < b.revision }
}
